package.json

```diff
@@ -770,6 +770,16 @@
               "lsp"
             ]
           },
+          "calva.replOutputThrottleRate": {
+            "markdownDescription": "If the repl outputs too quickly then results will be dropped from the output window. Setting this to 0 will disable throttling.",
+            "type": "number",
+            "default": 100
+          },
+          "calva.replOutputMaxLines": {
+            "markdownDescription": "The maximum number of lines to retain in the repl output window. Having the repl output window grow too large can significantly affect performance. Setting this to 0 will disable truncating",
+            "type": "number",
+            "default": 1000
+          },
```

Comment:
It's generally not the number of lines that is the problem, rather the number of tokens on the top level, or in a sexpr. A way to count tokens on the top level is to count words when printing output. That will introduce state that can be tricky to manage, though...

1000 lines is very low. It could be some 50 evaluations of not too big structures. This would be completely non-problematic for Calva to handle. While printing 1000 lines of output with 100 words each might start to be problematic (I actually don't know where the problems start.)

Reply:
Unlikely, maybe, but I know of a user who does this all the time. 😄  I search the file rather than scroll, but anyway. (I keep tons of browser tabs open as well.) To me it makes sense to disable this limit feature (by setting limit to 0, e.g.) and to have that as default.

Reply:
Hey @julienvincent, thanks for taking time to look into this.

I think deleting at an arbitrary line length will break syntax highlighting and maybe also formatting, unless I'm missing something. If I'm not though, I don't think that's really an option.

Reply:
> To me it makes sense to disable this limit feature (by setting limit to 0, e.g.) and to have that as default.

I'm happy with that too.

> I think deleting at an arbitrary line length will cause unbalanced forms and break syntax highlighting ...

It does break highlighting but in a localised way (only the broken form will be colored incorrectly). The rest of the file maintains the correct highlighting. I think this is an probably an acceptable behaviour.

> and maybe also formatting

Which formatting do you mean? I haven't noticed anything obviously wrong.

Reply:
![image](https://user-images.githubusercontent.com/8571121/212131380-a4e2ddb1-ef4d-4000-9f64-8392ff3292f9.png)

Here can see what I mean

Reply:
The way the structural editor works, when there is broken balance in the document, all bets are off.  It will sometimes seem to work (or even work), but in many cases the structural editor is broken and with that a lot of Calva functionality that depends on this.

If we limit on lines, it will have to do it such that the balance is never broken. There are some ways to do this, I can think of two right now:

1. Add the missing brackets. There is code for that in Calva, even if it might be a bit tricky to re-use (or, messy code, tbh).
2. Always print full results. The limit then need to happen before the buffering, I think. (Not that I know how it is done now, but anyway.)

That said, I think we could also consider popping up a warning when the line limit is reached, instead of enforcing the limit. The warning can have a button for clearing the window.

Reply:
> It does break highlighting but in a localised way (only the broken form will be colored incorrectly). The rest of the file maintains the correct highlighting. I think this is an probably an acceptable behaviour.

Yeah, I'd say that, in particular, is acceptable too.

By formatting I meant if someone is typing code in the repl window, the auto-formatting could/would be broken is there's an unbalanced form (which Peter mentioned above - structural editing). And as he said other functionality may be broken too in that case.